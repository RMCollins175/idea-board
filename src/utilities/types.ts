export interface IdeaType {
  title: string;
  description: string;
  timestamp: number;
  id: string;
}
