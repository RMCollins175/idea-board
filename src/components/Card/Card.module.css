.ideaContainer {
    min-width: 200px;
    max-width: 200px;
    min-height: 200px;
    max-height: 200px;
    background-color: aliceblue;
    border-radius: 10px;
    border: 1px solid black;
}

.ideaForm {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    padding: 10px;
}

.ideaTitleLabel, .ideaDescriptionLabel {
    display: flex; 
    border: 1px solid #eaeaea;
    border-radius: 4px;
    font-size: 12px;
}

.ideaTitleLabel {
    justify-content: space-between;
    align-items: center;
}

.ideaTitleInput {
    width: 70%
}

.ideaDescriptionLabel { 
    flex-direction: column;
    justify-content: flex-start; 
}

.ideaDescriptionInput {
    margin-top: 5px;
    align-self: center;
    max-width: 95%;
    max-height: 65px;
    min-height: 65px;
}

.ideaTimestamp {
    font-size: 12px;
    color: #a9a9a9;
}

.callToActionContainer {
    display: flex;
    justify-content: space-evenly;
    padding: 10px;
}