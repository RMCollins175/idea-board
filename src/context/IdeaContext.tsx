import React, { useEffect, useState } from "react";
import { useLocalStorage } from "../hooks/useLocalStorage";
import { IdeaType } from "../utilities/types";
import { v4 as uuidv4 } from "uuid";

export interface IdeaContextType {
  ideaTitle?: string;
  ideaDescription?: string;
  id?: string;
  timestamp?: number;
  ideas: IdeaType[];
  addIdea: (title: string, description: string) => void;
  updateIdea: (id: string, title: string, description: string) => void;
  deleteIdea: (id: string) => void;
}

export const IdeaContext = React.createContext<IdeaContextType>({
  ideas: [],
  addIdea: () => {},
  updateIdea: () => {},
  deleteIdea: () => {}
});

export const IdeaContextProvider = ({ children }: any) => {
  const [ideas, setIdeas] = useState<IdeaType[]>([]);
  const [ideasStorage, setIdeasStorage] = useLocalStorage("ideas", "[]");

  useEffect(() => {
    if (ideasStorage !== "[]") {
      setIdeas(JSON.parse(ideasStorage));
    }
  }, [ideasStorage]);

  const addIdea = (title: string, description: string) => {
    const newIdeas = [
      ...ideas,
      { title, description, timestamp: Date.now(), id: uuidv4() }
    ];
    setIdeas(newIdeas);
    setIdeasStorage(JSON.stringify(newIdeas));
  };

  const updateIdea = (id: string, title: string, description: string) => {
    const updatedIdeas = ideas.map((idea) => {
      if (idea.id === id) {
        return { ...idea, title, description, timestamp: Date.now() };
      }
      return idea;
    });
    setIdeas(updatedIdeas);
    setIdeasStorage(JSON.stringify(updatedIdeas));
  };

  const deleteIdea = (id: string) => {
    const filteredIdeas = ideas.filter((idea) => idea.id !== id);
    setIdeas(filteredIdeas);
    setIdeasStorage(JSON.stringify(filteredIdeas));
  };

  return (
    <IdeaContext.Provider value={{ ideas, addIdea, updateIdea, deleteIdea }}>
      {children}
    </IdeaContext.Provider>
  );
};
