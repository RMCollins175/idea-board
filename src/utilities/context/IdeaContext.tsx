import React from "react";
import { useIdeaHook } from "../hooks/useIdea";
import { IdeaType } from "../types";

export interface IdeaContextType {
  ideaTitle?: string;
  ideaDescription?: string;
  id?: number;
  timestamp?: number;
  ideas: IdeaType[];
  addIdea: (title: string, description: string) => void;
  updateIdea: (id: number, title: string, description: string) => void;
  deleteIdea: (id: number) => void;
}

// export const IdeaContext = React.createContext<IdeaContextType>({});
export const IdeaContext = React.createContext({});

export const IdeaContextProvider = ({ children }: any) => {
  const { ideas, addIdea, updateIdea, deleteIdea } = useIdeaHook();

  const IdeaContextValue: IdeaContextType = {
    ideas,
    addIdea,
    updateIdea,
    deleteIdea
  };

  return (
    <IdeaContext.Provider value={IdeaContextValue}>
      {children}
    </IdeaContext.Provider>
  );
};
