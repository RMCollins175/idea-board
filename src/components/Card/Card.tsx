import { useState, useContext } from "react";
import styles from "./Card.module.css";
import { format } from "date-fns";
import { IdeaContext } from "../../context/IdeaContext";

export interface CardProps {
  ideaTitle?: string;
  ideaDescription?: string;
  id?: string;
  timestamp?: number;
}

export const Card = ({
  ideaTitle,
  ideaDescription,
  id,
  timestamp
}: CardProps) => {
  const { addIdea, updateIdea, deleteIdea } = useContext(IdeaContext);

  const [inputTitle, setInputTitle] = useState(ideaTitle || "");
  const [inputDescription, setInputDescription] = useState(
    ideaDescription || ""
  );

  const handleSubmission = (event: any) => {
    event.preventDefault();
    if (id) {
      updateIdea(id, inputTitle, inputDescription);
    } else {
      if (addIdea) addIdea(inputTitle, inputDescription);
      setInputTitle("");
      setInputDescription("");
    }
  };

  const handleDelete = () => {
    if (id) {
      deleteIdea(id);
    }
  };

  const handleDescriptionChange = (event: any) => {
    if (event.target.value.length < 140) {
      setInputDescription(event.target.value);
    }
  };

  return (
    <div className={styles.ideaContainer}>
      <form onSubmit={handleSubmission} className={styles.ideaForm}>
        <label className={styles.ideaTitleLabel}>
          <input
            data-testid="IdeaForm.title"
            type="text"
            name="title"
            id="title"
            className={styles.ideaTitleInput}
            placeholder="Title"
            onChange={(event) => setInputTitle(event.target.value)}
            value={inputTitle}
            required
          />
        </label>
        <label
          className={styles.ideaDescriptionLabel}
          data-testid="IdeaForm.description"
        >
          <textarea
            name="description"
            className={styles.ideaDescriptionInput}
            maxLength={140}
            placeholder="Description"
            onChange={handleDescriptionChange}
            value={inputDescription}
          />
        </label>
        {id && (
          <p className={styles.ideaTimestamp}>
            {`${
              timestamp && format(new Date(timestamp), "yyyy-MM-dd - HH:mm:ss")
            }`}
          </p>
        )}
        <div className={styles.callToActionContainer}>
          <button
            type="submit"
            data-testid="IdeaForm.buttonAdd"
            disabled={!inputTitle || !inputTitle.trim()}
          >
            {id ? `Update` : `Add`}
          </button>
          <button
            className="ideaFormButton"
            data-testid="IdeaForm.buttonReset"
            onClick={() => {
              setInputTitle("");
              setInputDescription("");
            }}
          >
            Reset
          </button>
          {id && (
            <button data-testid="IdeaForm.buttonDelete" onClick={handleDelete}>
              Delete
            </button>
          )}
        </div>
      </form>
    </div>
  );
};
