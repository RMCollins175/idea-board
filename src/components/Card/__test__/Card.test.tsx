import { Card } from "../Card";
import renderer from "react-test-renderer";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { IdeaType } from "../../../utilities/types";
import { IdeaContext } from "../../../context/IdeaContext";
import { format } from "date-fns";

describe("Card component tests", () => {
  let idea: IdeaType = {
    id: "1",
    title: "Example idea",
    description: "Example description",
    timestamp: Date.now()
  };

  const dispatch = jest.fn();
  const ideas = [idea];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render the form with the correct inputs and labels", () => {
    render(
      <IdeaContext.Provider value={{ dispatch, ideas }}>
        <Card idea={idea} />
      </IdeaContext.Provider>
    );

    const titleInput = screen.getByPlaceholderText("Title");
    const descriptionInput = screen.getByPlaceholderText("Description");

    expect(titleInput).toBeInTheDocument();
    expect(descriptionInput).toBeInTheDocument();
  });

  it("should display the timestamp with the correct format", () => {
    const { getByText } = render(
      <IdeaContext.Provider value={{ dispatch, ideas }}>
        <Card idea={idea} />
      </IdeaContext.Provider>
    );
    const timestamp = getByText(
      format(new Date(idea.timestamp), "yyyy-MM-dd - HH:mm:ss")
    );
    expect(timestamp).toBeInTheDocument();
  });

  it("should render the 'Add' button when no idea is passed as a prop", () => {
    render(<Card />);
    const addButton = screen.getByText("Add");
    expect(addButton).toBeInTheDocument();
  });

  it("should have a 'Delete' button and be able to delete an idea when clicked", async () => {
    const { getByText } = render(
      <IdeaContext.Provider value={{ dispatch, ideas }}>
        <Card idea={idea} />
      </IdeaContext.Provider>
    );

    const deleteButton = getByText("Delete");
    expect(deleteButton).toBeInTheDocument();

    fireEvent.click(deleteButton);
    await waitFor(() => {
      expect(dispatch).toHaveBeenCalledWith({
        type: "DELETE_IDEA",
        payload: {
          id: idea.id
        }
      });
    });
  });

  it("should have a 'Reset' button and be able to reset the form when clicked", async () => {
    const { getByText } = render(
      <IdeaContext.Provider value={{ dispatch, ideas }}>
        <Card />
      </IdeaContext.Provider>
    );

    const resetButton = getByText("Reset");
    expect(resetButton).toBeInTheDocument();

    fireEvent.change(screen.getByPlaceholderText("Title"), {
      target: { value: "Test Title" }
    });
    fireEvent.change(screen.getByPlaceholderText("Description"), {
      target: { value: "Test Description" }
    });

    expect(screen.getByPlaceholderText("Title")).toHaveValue("Test Title");
    expect(screen.getByPlaceholderText("Description")).toHaveValue(
      "Test Description"
    );

    fireEvent.click(resetButton);

    await waitFor(() => {
      expect(screen.getByPlaceholderText("Title")).toHaveValue("");
      expect(screen.getByPlaceholderText("Description")).toHaveValue("");
    });
  });

  it("should submit the form, call handleSubmission function and update card with the correct data", async () => {
    render(
      <IdeaContext.Provider value={{ dispatch, ideas }}>
        <Card idea={idea} />
      </IdeaContext.Provider>
    );

    const titleInput = screen.getByPlaceholderText("Title");
    const descriptionInput = screen.getByPlaceholderText("Description");
    const submitButton = screen.getByText("Update");

    fireEvent.change(titleInput, { target: { value: "New title" } });
    fireEvent.change(descriptionInput, {
      target: { value: "New description" }
    });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(dispatch).toHaveBeenCalledWith({
        type: "UPDATE_IDEA",
        payload: {
          id: idea.id,
          title: "New title",
          description: "New description"
        }
      });
    });
  });

  it("should render a default Idea component form", () => {
    const CardForm = renderer.create(<Card />).toJSON();
    expect(CardForm).toMatchSnapshot();
  });
});
