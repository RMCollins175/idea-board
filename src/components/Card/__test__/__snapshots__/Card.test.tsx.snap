// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component tests snapshot test of Card 1`] = `
<div>
  <div
    class="ideaContainer"
  >
    <form
      class="ideaForm"
    >
      <label
        class="ideaTitleLabel"
      >
        <input
          class="ideaTitleInput"
          id="title-undefined"
          name="title"
          placeholder="Title"
          required=""
          value=""
        />
      </label>
      <label
        class="ideaDescriptionLabel"
      >
        <textarea
          class="ideaDescriptionInput"
          id="description-undefined"
          maxlength="140"
          name="description"
          placeholder="Description"
        />
      </label>
      <div
        class="callToActionContainer"
      >
        <button
          aria-label="Add Button"
          type="submit"
        >
          Add
        </button>
        <button
          type="button"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Card component tests snapshot test of Card with ideas 1`] = `
<DocumentFragment>
  <div
    class="ideaContainer"
  >
    <form
      class="ideaForm"
    >
      <label
        class="ideaTitleLabel"
      >
        <input
          aria-label="Example idea"
          class="ideaTitleInput"
          id="title-1"
          name="title"
          placeholder="Title"
          required=""
          value="Example idea"
        />
      </label>
      <label
        class="ideaDescriptionLabel"
      >
        <textarea
          aria-label="Example description"
          class="ideaDescriptionInput"
          id="description-1"
          maxlength="140"
          name="description"
          placeholder="Description"
        >
          Example description
        </textarea>
      </label>
      <p
        aria-label="Timestamp"
        class="ideaTimestamp"
      >
        2023-02-07 - 09:18:18
      </p>
      <div
        class="callToActionContainer"
      >
        <button
          aria-label="Update Button"
          type="submit"
        >
          Update
        </button>
        <button
          type="button"
        >
          Reset
        </button>
        <button
          aria-label="Delete Button"
        >
          Delete
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
